jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/AutoDemandingMessageHandler.java

```diff
@@ -0,0 +1,183 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.util;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.Utf8Appendable;
+import org.eclipse.jetty.util.Utf8StringBuilder;
+import org.eclipse.jetty.websocket.core.CloseStatus;
+import org.eclipse.jetty.websocket.core.Frame;
+import org.eclipse.jetty.websocket.core.FrameHandler;
+import org.eclipse.jetty.websocket.core.exception.BadPayloadException;
+import org.eclipse.jetty.websocket.core.exception.MessageTooLargeException;
+
+/**
+ * <p>
+ * A utility implementation of FrameHandler that de-fragments text frames and binary frames into a whole messages before
+ * calling {@link #onText(String, Callback)} or {@link #onBinary(ByteBuffer, Callback)}.
+ * </p>
+ * <p>
+ * This is a demanding frame handler, meaning {@link FrameHandler#isDemanding()} returns false.
+ * This means that flow control is by automatic when the callback is succeeded.
+ * It is preferable to use {@link DemandingMessageHandler} as the aggregation for binary messages can be done more
+ * efficiently if the frame handler can manage its own demand.
+ * </p>
+ * @see DemandingMessageHandler
+ */
+public class AutoDemandingMessageHandler extends AbstractMessageHandler
+{
+    private Utf8StringBuilder _textMessageBuffer;
+    private ByteArrayOutputStream _binaryMessageBuffer;
```

Comment:
`ByteBufferAccumulator` does copy the buffer, so you can use it instead of `ByteArrayOutputStream`.